dipy/reconst/csdeconv.py

```diff
@@ -798,3 +839,114 @@ def auto_response(gtab, data, roi_center=None, roi_radius=10, fa_thr=0.7,
         return response, ratio, indices[0].size
 
     return response, ratio
+
+
+def recursive_response(gtab, data, mask=None, sh_order=8, peak_thr=0.01,
+                       init_fa=0.08, init_trace=0.0021, iter=8,
+                       convergence=0.001, parallel=True,
+                       sphere=default_sphere):
+    """ Recursive calibration of response function using peak threshold
+
+    Parameters
+    ----------
+    gtab : GradientTable
+    data : ndarray
+        diffusion data
+    mask : ndarray
+        mask for recursive calibration, for example a white matter mask. It has
+        shape `data.shape[0:3]` and dtype=bool.
+    sh_order : int
+        maximal spherical harmonics order
+    peak_thr : float
+        peak threshold, how large the second peak can be relative to the first
+        peak in order to call it a single fiber population [1]
+    init_fa : float
+        FA of the initial 'fat' response function (tensor)
+    init_trace : float
+        trace of the initial 'fat' response function (tensor)
+    iter : int
+        maximum number of iterations for calibration
+    convergence : float
+        convergence criterion, maximum relative change of SH coefficients
+    sphere : Sphere
+        The sphere used for peak finding.
+
+    Returns
+    -------
+    response : ndarray
+        response function in SH coefficients
+
+    Notes
+    -----
+    In CSD there is an important pre-processing step: the estimation of the
+    fiber response function. Using an FA threshold is not a very robust method.
+    It is dependent on the dataset (non-informed used subjectivity), and still
+    depends on the diffusion tensor (FA and first eigenvector),
+    which has low accuracy at high b-value. This function recursively
+    calibrates the response function, for more information see [1].
+
+    References
+    ----------
+    .. [1] Tax, C.M.W., et al. NeuroImage 2014. Recursive calibration of
+           the fiber response function for spherical deconvolution of
+           diffusion MRI data.
+    """
+    S0 = 1
+    evals = fa_trace_to_lambdas(init_fa, init_trace)
+    res_obj = (evals, S0)
+
+    if mask is None:
+        data = data.reshape(-1, data.shape[-1])
+    else:
+        data = data[mask]
+
+    n = np.arange(0, sh_order + 1, 2)
+    where_dwi = lazy_index(~gtab.b0s_mask)
+    response_p = np.ones(len(n))
+
+    for num_it in range(1, iter):
+        r_sh_all = np.zeros(len(n))
+        csd_model = ConstrainedSphericalDeconvModel(gtab, res_obj,
```

Comment:
That is a fair point. The reason why I added this here, is that you ideally want to mimic a bit what is present in a real brain, which means that we probably have to make a (significantly) larger test dataset and that the test will take longer. But I defenitely agree that this would be beneficial.